crypto/fipsmodule/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,623 @@
+                                   const BN_ULONG *exp[2], const BN_ULONG *m,
+                                   const BN_ULONG *rr, const BN_ULONG k0[2],
+                                   int modulus_bitsize);
+
+/*
+ * Dual Montgomery modular exponentiation using prime moduli of the
+ * same bit size, optimized with AVX512 ISA.
+ *
+ * Input and output parameters for each exponentiation are independent and
+ * denoted here by index |i|, i = 1..2.
+ *
+ * Input and output are all in regular 2^64 radix.
+ *
+ * Each moduli shall be |factor_size| bit size.
+ *
+ * Supported cases:
+ *   - 2x1024
+ *   - 2x1536
+ *   - 2x2048
+ *
+ *  [out] res|i|      - result of modular exponentiation: array of qword values
+ *                      in regular (2^64) radix. Size of array shall be enough
+ *                      to hold |factor_size| bits.
+ *  [in]  base|i|     - base
+ *  [in]  exp|i|      - exponent
+ *  [in]  m|i|        - moduli
+ *  [in]  rr|i|       - Montgomery parameter RR = R^2 mod m|i|
+ *  [in]  k0_|i|      - Montgomery parameter k0 = -1/m|i| mod 2^64
+ *  [in]  factor_size - moduli bit size
+ *
+ * \return 0 in case of failure,
+ *         1 in case of success.
+ */
+int ossl_rsaz_mod_exp_avx512_x2(BN_ULONG *res1,
+                                const BN_ULONG *base1,
+                                const BN_ULONG *exp1,
+                                const BN_ULONG *m1,
+                                const BN_ULONG *rr1,
+                                BN_ULONG k0_1,
+                                BN_ULONG *res2,
+                                const BN_ULONG *base2,
+                                const BN_ULONG *exp2,
+                                const BN_ULONG *m2,
+                                const BN_ULONG *rr2,
+                                BN_ULONG k0_2,
+                                int factor_size)
+{
+    typedef void (*AMM)(BN_ULONG *res, const BN_ULONG *a,
+                        const BN_ULONG *b, const BN_ULONG *m, BN_ULONG k0);
+    int ret = 0;
+
+    /*
+     * Number of word-size (BN_ULONG) digits to store exponent in redundant
+     * representation.
+     */
+    int exp_digits = number_of_digits(factor_size + 2, DIGIT_SIZE);
+    int coeff_pow = 4 * (DIGIT_SIZE * exp_digits - factor_size);
+
+    /*  Number of YMM registers required to store exponent's digits */
+    int ymm_regs_num = NUMBER_OF_REGISTERS(exp_digits, 256 /* ymm bit size */);
+    /* Capacity of the register set (in qwords) to store exponent */
+    int regs_capacity = ymm_regs_num * 4;
+
+    BN_ULONG *base1_red, *m1_red, *rr1_red;
+    BN_ULONG *base2_red, *m2_red, *rr2_red;
+    BN_ULONG *coeff_red;
+    BN_ULONG *storage = NULL;
+    BN_ULONG *storage_aligned = NULL;
+    int storage_len_bytes = 7 * regs_capacity * sizeof(BN_ULONG)
+                           + 64 /* alignment */;
+
+    const BN_ULONG *exp[2] = {0};
+    BN_ULONG k0[2] = {0};
+    /* AMM = Almost Montgomery Multiplication */
+    AMM amm = NULL;
```

Comment:
can we get rid of this hack by implementing a wrapper AMM function that calls the correct implementation based on the input size?

Reply:
So, I started down this road with something like
```c
    if (!amm(factor_size, rr1_red, rr1_red, rr1_red, m1_red, k0_1)) {
	goto err;
    } /* (2) for m1 */
    if (!amm(rr1_red, rr1_red, coeff_red, m1_red, k0_1)) {
	goto err;
    } /* (3) for m1 */
```
Where `amm` looks like:
```c
int amm(int factor_size, BN_ULONG *res, const BN_ULONG *a,
        const BN_ULONG *b, const BN_ULONG *m, BN_ULONG k0) {
    switch (factor_size) {
    case 1024:
        rsaz_amm52x20_x1_ifma256(res, a, b, m, k0);
        break;
    case 1536:
        rsaz_amm52x30_x1_ifma256(res, a, b, m, k0);
        break;
    case 2048:
        rsaz_amm52x40_x1_ifma256(res, a, b, m, k0);
        break;
    default:
        return 0;
    }
    return 1
}
```

Which leaves us checking for error on each `amm` call. I think this may be a rare case where this otherwise ugly hack is a better option. WDYT?

Reply:
I think you can make this function take a pointer to the appropriate assembly function since you're providing it with `modlen` (factor_size) as well.

Reply:
True, but that feels leaky to me. The more I look at this, the less I agree that this is particularly ugly. What's here doesn't ask the caller to know the name of the assembly functions or need to intercept errors. I'm leaning more towards leaving it as it is.